### ISSUE TITLE
ParMesh::ReorientTetMesh fails after GeneralRefinement

### ISSUE THREAD

**Issue by sshiraiwa** [mfem/mfem#1028]
I want to refine a particular domain flagged by the domain attribute, keeping
the rest of mesh the same. I checked attribute array and make an Array<int> 
to list the element number I want to refine, and then I called GeneralRefinement. 

This works fine in serial but in parallel, I ran into the problem of getting "inconsistent 
vertex ordering found", when I call ReorientTetMesh after refinement. This error
does not happen always, but it tends to happen when I increase the number of refinement
(the number of loop count below), perhaps it is due to how the refinement is performed.
For your information, I attached a code (in Python) to show this problem, a mesh file,
and a screen capture to show how the mesh looks like when it works.
I really need to refine after ParMesh is generated. It would be great if you can
make suggestion to solve my problem. Thank you.

import numpy as np
import mfem.par as mfem
from mpi4py import MPI
num_proc = MPI.COMM_WORLD.size
myid     = MPI.COMM_WORLD.rank

smesh =  mfem.Mesh('box3d.msh', 1, 1, True)
pmesh =mfem.ParMesh(MPI.COMM_WORLD, smesh)

domains = [1]
for i in range(3): ### <<<--- refinement upto 2 did not make error.
    attr = pmesh.GetAttributeArray()

    # this collect index of element whose attr is 1
    idx = mfem.intArray(list(np.where(np.in1d(attr, domains))[0]))
    
    pmesh.GeneralRefinement(idx)

pmesh.ReorientTetMesh() ## <<< --- fails

Please change the file extension to .msh 
[box3d.msh.txt](https://github.com/mfem/mfem/files/3477672/box3d.msh.txt)
![refined_mesh](https://user-images.githubusercontent.com/1929159/62635287-67dca800-b905-11e9-8bad-2b51d5b22c5e.png)

**Comment by mlstowell** [comment]
Hi, @sshiraiwa ,

I'm not sure if it will matter but I want to be sure I get this correct.  Is the line `pmesh.ReorderTetMesh()` inside the `for` loop or after it?

Best wishes,
Mark

**Comment by sshiraiwa** [comment]
ReorientTetMesh is outside the loop, It called once after GeneralRefinement is called several times.

**Comment by mlstowell** [comment]
Thanks, Syun'ichi,

I've been able to reproduce the problem.  I'll try to take a closer look after lunch and get back to you.

Best wishes,
Mark

**Comment by v-dobrev** [comment]
One thing, there is a missing comma in: `mfem.Mesh('box3d.msh', 1, 1 True)` -- could this be the issue?

Also, you can try calling `pmesh.Finalize(true);` before the local refinement loop. This is needed if you perform uniform refinement before local refinement -- not sure if it is also needed in your case.

**Comment by sshiraiwa** [comment]
Hi Veselin

I fixed the typo in the code segment ;D I am not sure either if  Finalize(True) is needed.
As far as I did try, I didn't do anything.

--
S

**Comment by mlstowell** [comment]
I tried adding `pmesh->Finalize(true)` in three different places and didn't see any change in the error.

**Comment by mlstowell** [comment]
Hi, @sshiraiwa ,

I've been looking at this most of the day and I haven't found an explanation.  Currently I'm wondering if there might be a subtle bug in the red-green refinement algorithm.  @v-dobrev and @jakubcerveny, do you happen to know of a reference paper or other text that describes this algorithm?  I expect the tetrahedron types PU, A, PF, O, and M must be described somewhere and I'm hoping that if I read that reference I might understand the algorithm better.

Thanks,
Mark

**Comment by jakubcerveny** [comment]
I believe the algorithm comes from this paper: Arnold et al.: Locally adapted tetrahedral meshes using bisection (2000), http://www-users.math.umn.edu/~arnold/papers/bistet.pdf

**Comment by v-dobrev** [comment]
One thing that can be tried is to do the refinements in serial then create the `ParMesh` and then call `pmesh.ReorientTetMesh()` -- just to see if the issue is specific to the parallel implementation of refinement.

**Comment by mlstowell** [comment]
@v-dobrev, it seems to work fine in serial.  

It looks like we never posted the error message.  Here it is:
```
Verification failed: (stria_flag[i] == stria_master_flag[i]) is false:
 --> inconsistent vertex ordering found
 ... in function: virtual void mfem::ParMesh::ReorientTetMesh()
 ... in file: mesh/pmesh.cpp:2449
```
The trouble seems to be that the lowest numbered vertex does not occur in the same location on some shared faces.

**Comment by mlstowell** [comment]
Hi, @sshiraiwa ,
@v-dobrev and I spoke about this issue this afternoon.  We've each been looking into it and the conclusion is that this is a non-trivial bug in the local tetrahedral refinement algorithm.  We'll continue to try to find a solution but in the mean time @v-dobrev came up with a clever work around (which you may have already found).  Essentially the trick is not to call `Mesh::ReorientTetMesh`.  This should work as long as you're not using high order Nedelec basis functions (stick for order 1 for example).

Best wishes,
Mark

**Comment by sshiraiwa** [comment]
Dear Mark and Veselin

Thank you for looking into this, and the interesting trick, which I haven't thought about.  I will try this. Also, please let me know when you have a time to fix the algorithmic issue. Thank you.

Best
Syun'ichi

**Comment by mlstowell** [comment]
Hi, @jakubcerveny and @v-dobrev ,

I think I found a bug in the `Mesh::Bisection` algorithm.  However, I'm hesitant to start tinkering with it because I don't fully understand some of its functionality (such as the variables `tr1` and `tr2`).

Consider the switch statements on the following lines:
https://github.com/mfem/mfem/blob/18f0cf0873d9f5c941c747e73bbdd951bba7444c/mesh/mesh.cpp#L7480-L7509
I believe these are supposed to check for marked edges on the two faces that are not being cut during this bisection.  However, edges 2, 3, and 5, in the first switch statement, are not on a common face nor are edges 1, 4, and 5 from the second switch statement.  I think the first switch statement should involve edges 1, 2, and 5 while the second involves 3, 4, and 5.

For reference, I'm using the edge indices implied by:
 https://github.com/mfem/mfem/blob/18f0cf0873d9f5c941c747e73bbdd951bba7444c/fem/geom.cpp#L806-L807

Have I misinterpreted something or does this look like a genuine problem?

Best wishes,
Mark

**Comment by v-dobrev** [comment]
Mark,

class `Tetrahedron` seems to use a different local numbering for the 6 edges, see e.g.
https://github.com/mfem/mfem/blob/3f1adff17d8865061ef643cabb11d7a49c5fd53c/mesh/tetrahedron.cpp#L225-L234

In the above code, `ind[0]` and `ind[1]` store the refinement edges for the faces with vertex indices `(0,2,3)` and `(1,2,3)`, respectively. For example, edge with vertices `(0,2)` is stored as edge `2`, `(0,3)` as `3`, etc.

The two indices in `ind` are stored in `Tetrahedron::refinement_flag` by the call https://github.com/mfem/mfem/blob/3f1adff17d8865061ef643cabb11d7a49c5fd53c/mesh/tetrahedron.cpp#L277

These two numbers are then extracted back into `old_redges` by the call https://github.com/mfem/mfem/blob/18f0cf0873d9f5c941c747e73bbdd951bba7444c/mesh/mesh.cpp#L7471

Thus, the code in `Mesh::Bisection` that you linked above does not use the edge indices as defined in `geom.cpp` -- instead it uses the enumeration defined by class `Tetrahedron`.

**Comment by mlstowell** [comment]
Hi, @v-dobrev (and welcome back!),

I might argue that this is another manifestation of the same bug.  The `Tetrahedron::GetEdgeVertices()` method uses the edges defined in `geom.cpp` through its `typedef Geometry::Constants<Geometry::TETRAHEDRON> geom_t;` member data.

Clearly, it seems that we have two different edge numbering schemes for tetrahedra.  It's possible that we have two subsets of methods which are each internally consistent but don't agree with each other.  This could lead to some subtle errors that would be quite difficult to notice.

Shall we try to standardize on the edge numbering scheme defined in `geom.cpp`?

Best wishes,
Mark

**Comment by v-dobrev** [comment]
Yes, you are right, this different edge numbering is specific to methods for dealing with the refinement flag.

Changing this edge enumeration to match the one in `geom.cpp` may be tricky and I don't see a compelling reason to do it. I think this is not relevant to the issue we see in parallel since everything works fine in serial. I believe that the issue is in the implementation of `ParMesh::ReorientTetMesh` and not with the refinement itself.

**Comment by mlstowell** [comment]
I agree.  The edge numbering is probably not related to this parallel issue.

**Comment by v-dobrev** [comment]
Hi @sshiraiwa,

I think we resolved this issue in PR #1119 (branch `uniform-refinement-element-ordering-dev`) -- at least the test that I used to reproduce your issue works fine there.

**Comment by tzanio** [comment]
Discussion continued in #1119. @sshiraiwa, thanks for reporting this.
